### ISSUE TITLE
Call rtcIntersectNp in tbb::parallel_for

### ISSUE THREAD

**Issue by Speierers** [embree/embree#301]
Hi,

I would like to call `rtcIntersectNp` on multiple threads, where  every thread/call would compute the intersections for a subset of the rays. Something like this:

```c++
// Initialize pointer to the whole set of rays
RTCRayHitNp rh = ...;

// Could disable parallel "mode", which works fine
#if PARALLEL_LLVM_RAY_INTERSECT
    tbb::parallel_for(
        tbb::blocked_range<size_t>(0, N, 1),
            [&](const tbb::blocked_range<size_t> &range) {
                size_t offset = range.begin(),
                       size   = range.end() - offset;
                // Create an RTCRayHitNp where the pointers are offset to point to the first ray of the subset
                RTCRayHitNp rh_;
                rh_.ray = offset_rtc_ray(rh.ray, offset);
                rh_.hit = offset_rtc_hit(rh.hit, offset);

                // Only intersect the subset of rays
                rtcIntersectNp(s.accel, &context, &rh_, size);
            }
    );
#else
    rtcIntersectNp(s.accel, &context, &rh, N);
#endif
```

This "parallel" code works most of the time, but I sometimes get the following error:

```
embree/kernels/bvh/bvh_intersector_hybrid.cpp:159: static void embree::sse2::BVHNIntersectorKHybrid<4, 4, 1, false, embree::sse2::ArrayIntersectorK_1<4, embree::sse2::TriangleMIntersectorKMoeller<4, 4, 4, false> >, true>::intersect(vint<K> *__restrict, Accel::Intersectors *__restrict, RayHitK<K> &__restrict, embree::IntersectContext *__restrict) [N = 4, K = 4, types = 1, robust = false, PrimitiveIntersectorK = embree::sse2::ArrayIntersectorK_1<4, embree::sse2::TriangleMIntersectorKMoeller<4, 4, 4, false> >, single = true]: Assertion `all(valid, ray.valid())' failed.
Aborted (core dumped)
```

It is suppose to be safe to call `rtcIntersectNp` in a `parallel_for` loop?

Thanks!

**Comment by Speierers** [comment]
Looks like setting `EMBREE_IGNORE_INVALID_RAYS` to `ON` solves my issue. So maybe this is a bug?

**Comment by cbenthin** [comment]
The parallel code should work but it seems you are passing somehow invalid rays to Embree, otherwise the ignoring invalid rays flag would not help. Is there a chance to get your parallel code so that we can try to reproduce it on our side?

**Comment by Speierers** [comment]
I do have invalid rays. In our wavefront renderer, path might be killed before others (e.g. russian roulette). In which case I will set those rays as invalid: `ray.tfar > ray.tnear`. What would be the expected behavior when `EMBREE_IGNORE_INVALID_RAYS=OFF` in this case?

> Is there a chance to get your parallel code so that we can try to reproduce it on our side?

Not really, the code isn't ready to be shared at this point unfortunatelly.

**Comment by cbenthin** [comment]
Do you set rays as invalid by ray,tfar > ray,tnear? That makes them valid. It's the other way around.
Which version of Embree are you using and did you set any special ray flags, e.g. coherent?

As you are using the "structure of pointers" mode for the ray stream, it seems the relevant line is 585 in
https://github.com/embree/embree/blob/master/kernels/bvh/bvh_intersector_stream_filters.cpp

Here we deactivate rays that have tnear > tfar. Can you do a quick test and add a 
```
PRINT(ray);
assert(valid == (valid & ray.valid()));
```
thereby verify in the console output that rays that have a valid mask entry have valid data.
Thanks.

**Comment by Speierers** [comment]
> Do you set rays as invalid by ray,tfar > ray,tnear? That makes them valid. It's the other way around.

Indeed, what I meant is `ray.tfar < ray.tnear`.

I always set my `ray.mask` to `0` and `EMBREE_RAY_MASK` is `OFF`.

> Which version of Embree are you using 

This is with latest `master` (732ba86)

> did you set any special ray flags, e.g. coherent?

No flags set.

**Comment by svenwoop** [comment]
I think I know what may be going on here. This is a raise condition caused by a bug in our usage of SSE to store the ray results back again. We update the target array in SSE by performing an SSE load, modifying the vector, and storing it back again. However, this will not work when multiple threads are working on that same array as the store may corrupt data just written by the neighboring thread.

Could you please change your code to always pass blocks of 4 rays to Embree? I think changing tbb::blocked_range<size_t>(0, N, 1) to tbb::blocked_range<size_t>(0, N, 4) should do the job. This will workaround these raise conditions.

**Comment by Speierers** [comment]
Hi,

Very sorry for my late reply.

Setting `EMBREE_IGNORE_INVALID_RAYS` to `ON` solved my issue above but I was still getting some memory corruption (random segfault later on) when setting the granularity to `1`. I figured that a while ago (now setting it to `1024`) and it works perfecly fine.

@svenwoop thanks for your clarification. It's always good to understand where the problem was coming from even though I had found a work around!

Thanks for your help!
